### ISSUE TITLE
Painful to use edge cases with Execute.WithConnection

### ISSUE THREAD

**Issue by jzabroski** [fluentmigrator/fluentmigrator#987]
Consider the following situation:
1. Database default collation was incorrectly set-up as `Latin1_General_CI_AI` - it should be `SQL_Latin1_General_CP1_CI_AS`
2. All columns in the database are therefore currently collated as `Latin1_General_CI_AI`
3. You have a table-valued function that returns a table that contains an nvarchar(50) with no collation specified (as is typical... I don't think I've ever seen anyone specify collation on a programmability object like functions and stored procedures). - This indirectly causes schema binding to occur.
4. To change the collation of columns, you have to:
    1. Drop any objects that depend on schema binding for character data
    2. Alter collation at the database-level
    3. Re-create any objects that depended on schema binding for character data
    4. For each column in `INFORMATION_SCHEMA.COLUMNS` that is of data type NVARCHAR, VARCHAR, NCHAR or CHAR:
        1. Sort ascending whether the column belongs to an `INFORMATION_SCHEMA.TABLES` with `TABLE_TYPE` = 'BASE TABLE' or 'VIEW' (to ensure base tables are fixed before dependent objects like views are fixed).
        2. If the column belongs to an `INFORMATION_SCHEMA.TABLES` with `TABLE_TYPE` = 'BASE TABLE', check whether the column belongs to any constraints, indexes, primary keys, or foreign keys. If it does, generate the drop statement for those dependent objects.
        3. If the column belongs to an `INFORMATION_SCHEMA.TABLES` with `TABLE_TYPE` = 'BASE TABLE', then generate an ALTER TABLE ALTER COLUMN expression to change the collation.
        4. Re-apply any constraints, indexes, primary keys, or foreign keys you dropped in step 2.
        5. If the column belongs to an `INFORMATION_SCHEMA.TABLES` with `TABLE_TYPE` = 'VIEW', execute `sp_refreshview` for that view.

In FM, the solution, today, is pretty much to use Execute.WithConnection, due to limitations in T-SQL's Transactional DDL.  You might be able to write all of this in one giant dynamic sp_executesql statement, but I doubt it.  Even if it's possible, you would need a lot of nested sp_executesql statements to delineate batches of CREATE code.  The code gets even nastier to write and harder to read if the constraints use auto-generated names.

When using Execute.WithConnection, it will work, but you have to properly delineate the statements into batches.

I can, of course, propose a design, but I wonder if I'd be over-optimizing for SQL Server users. @fubar-coder  Do you have any statistics on how many people use SQL Server vs. DB2 vs. Oracle vs. Postgres vs. Firebird vs. SQLite, etc?  Do you use SQL Server at work or a different SQL DB technology?

**Comment by fubar-coder** [comment]
No, I don't have such statistics. What would your proposed solution be?

**Comment by jzabroski** [comment]
I haven't fully fleshed it out, but I think some form of Promises as a way to model future computations to be run would be quite effective.  In a sense, most of what FluentMigrator does can be described using Monad Comprehensions as well - and Execute.With is just a big hammer way to put a bunch of global state into a migration script.   With Monad Comprehensions, you just have to "thread the state" through the migration.  So, the downside is you have to know your initial state ahead of time.  This links back to the discussion we had about a generic IState abstraction in #955

For example, one common thing I need to do in Execute.WithConnection migrations is something like the following template:

```csharp
    [Migration(2019021104, TransactionBehavior.None, "happy valentine's day")]
    public class _2019021401_happy_valentines_day : ForwardOnlyMigration
    {
        public override void Up()
        {
            var databaseName = new SqlConnectionStringBuilder(ConnectionString).InitialCatalog;
            var cmdText1 = $@"SELECT '{databaseName}' AS BatchOne";
            var cmdText2 = $@"SELECT '{databaseName}' AS BatchTwo";
Execute.WithConnection((conn, tran) =>
            {
                using (var conn2 = new SqlConnection(conn.ConnectionString))
                {
                    using (var cmd = conn2.CreateCommand())
                    {
                        Console.WriteLine($"{cmdText1}");
                        cmd.CommandText = cmdText1;
                        conn2.Open();
                        cmd.ExecuteNonQuery();
                    }
                }

                Console.WriteLine("Starting Batch 2...");
                using (var conn2 = new SqlConnection(conn.ConnectionString))
                {
                    using (var cmd = conn2.CreateCommand())
                    {
                        Console.WriteLine($"{cmdText2}");
                        cmd.CommandText = cmdText2;
                        conn2.Open();
                        cmd.ExecuteNonQuery();
                    }
                }
        }
```

There's a couple of things here that strike me as odd:
1. Because I have TransactionBehavior.None, I have to make sure I create new connections via the conn lambda parameter, rather than the tran lambda parameter.  tran will be null
2. Because I have exited the graces of FM syntax, I have to know the names of any auto-generated objects, which, if I am rebuilding a development environment from scratch on every build, is a constantly moving target. If I could somehow thread the FM syntax into Execute.WithConnection or some similar "let me do some fancy SQL stuff" abstraction, that would save me from writing a bunch of home brewed extension methods on an ad-hoc basis.
3. The tool can't help me in knowing cmdText1 and cmdText2 need to be executed as separate batches, and so depending on what those commands are, I can get zombied sql transaction errors and have to dip into SQL Profiler and turn on "User Error Messages" to figure out what is causing the zombied transaction.  This makes both writing and debugging the code hard. - This is a subtle point that most engineers I explain this to do not grasp.

**Comment by fubar-coder** [comment]
@jzabroski Can you please create a new issue and - maybe - create a small prototype that shows what you'd like it to look like - covering API and use cases?

**Comment by jzabroski** [comment]
OK, I'll try, but it's probably at least 40 hours of work on my end.  Note, I am not trying to push this on you, just sharing my feedback of using your great tool and what I feel the present limitations are.

**Comment by fubar-coder** [comment]
Don't invest too much time. I just need to get a feel for what you meant/described.

**Comment by jzabroski** [comment]
Here is a contemporaneous note, which I will use to flesh out my idea further.

In functional programming, it's desirable to be able to describe side effects, like logging, but to do so in a way that doesn't break referential transparency and reasoning about program state.  To do so, we "thread" global state from the start of the computation all the way through the end of the computation.  Some people call this "Railway Oriented Programming".

Here is a subtle hint at the power of restructuring things as rails (monad comprehensions).

It's hard to reason about the effects of Execute.WithConnection, because:

1. Typically, Execute.Sql and other migrations automatically log the SQL as it's executed.  FM automatically adds, as a side-effect "threaded through the computation", log messages which are largely annotated ahead of time with [+] and other useful information.
```
-------------------------------------------------------------------------------
2018120302: DisableForeignKeys migrating
-------------------------------------------------------------------------------
[+] Beginning Transaction
[+] ExecuteSqlStatement ALTER TABLE dbo.OM NOCHECK CONSTRAINT FK_OM_OFId_F_FId
ALTER TABLE dbo.OtherManager NOCHECK CONSTRAINT FK_OM_OFId_F_FId
=> 0.024154s
[+] ExecuteSqlStatement ALTER TABLE dbo.FD NOCHECK CONSTRAINT FK_FD_SId_S_SId
ALTER TABLE dbo.FD NOCHECK CONSTRAINT FK_FD_SId_S_SId
=> 0.0281397s
INSERT INTO [Monitor].[VersionInfo] ([Version], [AppliedOn], [Description]) VALUES (2018120302, '2019-03-01T20:51:15', 'Disable FKs due to third party providing bad data.')
[+] Committing Transaction
[+] 2018120302: DisableForeignKeys migrated
=> 0.0775537s
```
To do the same thing with Execute.WithConnection, you have to manually write your own Console.WriteLine.  And now, if FM changes how it writes output, similar to xunit 2.0 changing how it handles output, Console.WriteLine will be a breaking change at some point.
2. You cannot weave Execute.Sql and Execute.WithConnection in the same migration in a top-down, readable style - the order of SQL actually executed is different from the order in which they may appear, since one is a Lambda and the other is internally a Builder pattern that builds a set of commands to execute once the Runner is ready to execute it.
3. You cannot weave TransactionBehavior through a computation.  It has to be declared ahead of time.  The only workaround for this is Execute.WithConnection and nasty hacks like I post above.  In case you did not see the hack above, note I am creating a new connection inside Execute.WithConnection.
